### PR TITLE
[api+sdk] Groundwork for cross-cutting builder extensions

### PR DESCRIPTION
**Pull request by CodeBlanch**
Relates to #5137

## Changes

* Adds `IOpenTelemetryBuilder` in API (extensions).
* Adds `IOpenTelemetryBuilder` extensions in SDK which now drive the "With" operations we have today.

## Details

The goal is to allow components to author extensions which target all signals (see #4940):

```csharp
public static IOpenTelemetryBuilder AddOtlpExporter(this IOpenTelemetryBuilder builder) {}
```

* We don't want to force things to take a dependency on `OpenTelemetry.Extensions.Hosting` to do this. This is why `IOpenTelemetryBuilder` has been introduced.
* We don't want to force things to take a dependency on SDK to do this. This is why `IOpenTelemetryBuilder` is going into `OpenTelemetry.Api.ProviderBuilderExtensions`.
* The "With" extensions in SDK are being added to handle "type switching" which will result anytime a cross-cutting extension is invoked:

   Today:
   ```csharp
   services.AddOpenTelemetry()
       .ConfigureResource(...) // invoke OpenTelemetryBuilder.ConfigureResource instance method
       .WithTracing(...) // invoke OpenTelemetryBuilder.WithTracing instance method
       .WithMetrics(...) // invoke OpenTelemetryBuilder.WithMetrics instance method
   ```

   Future:
   ```csharp
   services.AddOpenTelemetry()
       .AddOtlpExporter() // invoke extension on IOpenTelemetryBuilder and return IOpenTelemetryBuilder
       .ConfigureResource(...) // invoke IOpenTelemetryBuilder ConfigureResource SDK extension method
       .WithTracing(...) // invoke IOpenTelemetryBuilder WithTracing SDK extension method
       .WithMetrics(...) // invoke IOpenTelemetryBuilder WithMetrics SDK extension method
   ```

A secondary goal this also lays groundwork for is to offer a new bootstrap pattern for .NET Framework (really manual/non-host sdk management) which uses the "With" pattern our hosting package offers. This was explored a bit on #5137 and mentioned [here](https://github.com/open-telemetry/opentelemetry-dotnet/pull/5262#discussion_r1468170598). That pattern is more concise, allows the cross-cutting extensions to be used more widely, and (hopefully) simplifies our documentation.

## Public API Changes

* OpenTelemetry.Api.ProviderBuilderExtensions

```diff
namespace OpenTelemetry;

+public interface IOpenTelemetryBuilder
+{
+    IServiceCollection Services { get; }
+}
```

* OpenTelemetry

```diff
namespace OpenTelemetry;

+ public static class OpenTelemetryBuilderSdkExtensions
+ {
+     public static IOpenTelemetryBuilder ConfigureResource(this IOpenTelemetryBuilder builder, Action<ResourceBuilder> configure) {}
+     public static IOpenTelemetryBuilder WithMetrics(this IOpenTelemetryBuilder builder) {}
+     public static IOpenTelemetryBuilder WithMetrics(this IOpenTelemetryBuilder builder, Action<MeterProviderBuilder> configure) {}
+     public static IOpenTelemetryBuilder WithTracing(this IOpenTelemetryBuilder builder) {}
+     public static IOpenTelemetryBuilder WithTracing(this IOpenTelemetryBuilder builder, Action<TracerProviderBuilder> configure) {}

#if EXPOSE_EXPERIMENTAL_FEATURES
+     public static IOpenTelemetryBuilder WithLogging(this IOpenTelemetryBuilder builder) {}
+     public static IOpenTelemetryBuilder WithLogging(this IOpenTelemetryBuilder builder, Action<LoggerProviderBuilder> configure) {}
+     public static IOpenTelemetryBuilder WithLogging(this IOpenTelemetryBuilder builder, Action<LoggerProviderBuilder>? configureBuilder, Action<OpenTelemetryLoggerOptions>? configureOptions) {}
#endif
+ }
```

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [X] ~Unit tests added/updated~ There are existing unit tests in Hosting for the extensions being added to SDK. I decided to leave them there for now. They really should be moved into the SDK tests now but that is going to take some refactoring. I'll do that separately so as to not explode the diff.
* [X] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Review by rajkumar-rangaraj — APPROVED** [review]
LGTM.

File: src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs [hunk 1]
```diff
@@ -123,126 +109,7 @@ public OpenTelemetryBuilder WithTracing()
...
-    /// <remarks><inheritdoc cref="WithLogging()" path="/remarks"/></remarks>
-    /// <param name="configureBuilder">Optional <see
-    /// cref="LoggerProviderBuilder"/> configuration callback.</param>
-    /// <param name="configureOptions">Optional <see
-    /// cref="OpenTelemetryLoggerOptions"/> configuration callback.</param>
-    /// <returns>The supplied <see cref="OpenTelemetryBuilder"/> for chaining
-    /// calls.</returns>
-    internal
-#endif
-        OpenTelemetryBuilder WithLogging(
```
Comment (alanwest) [inline src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs:239]: What's the overall sequence for landing this work?

Could removing `WithLogging` methods in this PR be premature?

Without them there's no way to really use the new `WithLogging` methods that have moved to the SDK until additional work is landed, correct?

File: src/OpenTelemetry.Api/CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,10 @@
 
 ## Unreleased
 
+* Added `IOpenTelemetryBuilder` interface to support authoring extensions which
+  can configure multiple OpenTelemetry signals (tracing, metrics, and/or logs).
+  ([#5265](https://github.com/open-telemetry/opentelemetry-dotnet/pull/5265))
```
Comment (alanwest) [inline src/OpenTelemetry.Api/CHANGELOG.md:7]: I think this changelog entry was meant for OpenTelemetry.Api.Providerbuilders. But also, maybe no changelog is necessary yet? Is there really much an end user can do with `IOpenTelemetryBuilder` until more work is landed?

Comment (CodeBlanch) [reply src/OpenTelemetry.Api/CHANGELOG.md]: > I think this changelog entry was meant for OpenTelemetry.Api.Providerbuilders.

Fixed thanks!

> But also, maybe no changelog is necessary yet? Is there really much an end user can do with IOpenTelemetryBuilder until more work is landed?

Yes! After this PR users can use `IOpenTelemetryBuilder` to authoring cross-cutting things. Like...

```csharp
public static IOpenTelemetryBuilder AddOtlpExporter(this IOpenTelemetryBuilder builder)
{
    builder.Services.ConfigureOpenTelemetryTracerProvider(builder => builder.AddOtlpExporter());
    builder.Services.ConfigureOpenTelemetryMeterProvider(builder => builder.AddOtlpExporter());
    builder.Services.ConfigureOpenTelemetryLoggerProvider(builder => builder.AddOtlpExporter());

    return builder;
}
```

File: src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs (hunk 1 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs]: Had an offline conversation with @alanwest about this.

If you had before...

```csharp
services.AddOpenTelemetry().WithLogging();
```

...you now need to do...

```csharp
using OpenTelemetry;
services.AddOpenTelemetry().WithLogging();
```

For that reason we decided to put back the `WithLogging` methods on `OpenTelemetryBuilder` for now. We might do more work on follow-ups though.

**Review by alanwest — APPROVED** [review]

**Comment by cmeeren** [comment]
Could the migration path be clarified?

I currently have this, to set up telemetry with HotChocolate, as explained a few snippets down from [this heading](https://chillicream.com/docs/hotchocolate/v13/server/instrumentation#setup):

```f#
builder.Services
    .AddOpenTelemetry()
    .WithTracing(fun b ->
        b.AddHotChocolateInstrumentation().SetResourceBuilder(otlResourceBuilder)
        |> ignore

        let connStr =
            builder.Configuration.GetValue<string> "APPLICATIONINSIGHTS_CONNECTION_STRING"

        if not (String.isNullOrEmpty connStr) then
            b.AddAzureMonitorTraceExporter(fun o -> o.ConnectionString <- connStr) |> ignore
        |> ignore
    )
|> ignore
```

After updating `OpenTelemetry.Extensions.Hosting`, this now fails compilation (due to warnings as errors) because of the obsoletion. Based on the OP here, I don't understand what I can change to make this work.

**Comment by StachuDotNet** [comment]
Wondering the same thing as @cmeeren. Any advice would be appreciated, @CodeBlanch :)

**Comment by CodeBlanch** [comment]
@cmeeren @StachuDotNet This is interesting. It seems the same code in C# works fine but for some reason with F# it warns. I added this to discuss on our SIG meeting today. I'm considering just removing the obsoletion. I'll report back after our discussion.

**Pull request merged**